crates/subspace-farmer/src/plot.rs

```diff
@@ -485,21 +503,26 @@ impl IndexHashToOffsetDB {
 
     /// Returns `true` if piece plot will not result in exceeding plot size and doesn't exist
     /// already
-    fn should_store(&self, index_hash: &PieceIndexHash) -> io::Result<bool> {
-        Ok(match self.max_distance {
-            Some(max_distance) => {
-                PieceDistance::xor_distance(index_hash, &self.address) < max_distance
-                    && self.get(index_hash)?.is_none()
-            }
-            None => false,
-        })
+    fn should_store(&self, index_hash: &PieceIndexHash) -> bool {
+        let key = self.get_key(index_hash);
+        match self.key_bounds {
+            Some((lower, upper)) => key >= lower && key <= upper,
+            None => true,
+        }
     }
 
     fn remove_furthest(&mut self) -> io::Result<Option<PieceOffset>> {
-        let max_distance = match self.max_distance {
-            Some(max_distance) => max_distance,
+        let (lower_bound, upper_bound) = match self.key_bounds {
+            Some(key_bounds) => key_bounds,
             None => return Ok(None),
         };
+        let max_distance =
+            if (PieceDistance::MAX / 2) - lower_bound < upper_bound - (PieceDistance::MAX / 2) {
```

Comment:
I don't think `lower_bound` is guaranteed to be smaller than `PieceDistance::MAX / 2` and similarly `upper_bound` is not guaranteed to be bigger`. Also in such cases you really want to use checked, overflowing or wrapping math explicitly. I believe what you want here is to check unidirectional distance between mid point and upper/lower bound, even if they go on number line as `lower_bound, upper_bound, mid_point` or `mid_point, lower_bound, upper_bound`.

Some intermediate variables and named constants for things like `PieceDistance::MAX / 2` would go a long way here. Many will not try to visualize this math in their head and will instead think like "OK, looks good", so you really want to help them understand what is going on, like I tried to do in https://github.com/subspace/subspace/blob/f59154ae1cd7d2df9b233c3b77f108f1d82bc3d1/crates/sp-consensus-subspace/src/digests.rs#L40-L46